shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -20,60 +20,104 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
   type Raise = Diagnostic => Unit
   type Binding = Str -> TypeScheme
   type Bindings = Map[Str, TypeScheme]
-  
+
   case class TypeDef(
     kind: TypeDefKind,
     nme: TypeName,
-    tparams: List[TypeName],
-    targs: List[TypeVariable],
-    body: Type,
+    tparamsargs: List[(TypeName, TypeVariable)],
+    tvars: Iterable[TypeVariable], // "implicit" type variables. instantiate every time a `TypeRef` is expanded
+    bodyTy: SimpleType,
     mthDecls: List[MethodDef[Right[Term, Type]]],
     mthDefs: List[MethodDef[Left[Term, Type]]],
     baseClasses: Set[Var],
     toLoc: Opt[Loc],
   ) {
     def allBaseClasses(ctx: Ctx)(implicit traversed: Set[Var]): Set[Var] =
-      baseClasses.map(v => Var(v.name.toList.mapHead(_.toLower).mkString)) ++
+      baseClasses.map(v => Var(v.name.decapitalize)) ++
         baseClasses.iterator.filterNot(traversed).flatMap(v =>
           ctx.tyDefs.get(v.name).fold(Set.empty[Var])(_.allBaseClasses(ctx)(traversed + v)))
-    val targsMap: Map[Str, TypeVariable] = tparams.map(_.name).zip(targs).toMap
+    val (tparams: List[TypeName], targs: List[TypeVariable]) = tparamsargs.unzip
+    def thisTy(implicit prov: TypeProvenance): TypeRef = TypeRef(nme, targs)(prov)
+    def wrapMethod(pt: PolymorphicType)(implicit prov: TypeProvenance): MethodType =
+      MethodType(pt.level, S(FunctionType(singleTup(thisTy), pt.body)(prov)), nme)
+    def wrapMethod(mt: MethodType)(implicit prov: TypeProvenance): MethodType =
+      if (mt.body.nonEmpty) wrapMethod(mt.toPT) else mt.copy(parents = nme :: Nil)
+  }
+
+  private case class MethodDefs(
+    tn: TypeName,
+    parents: List[MethodDefs],
+    decls: Map[Str, MethodType],
+    defns: Map[Str, MethodType],
+  ) {
+    private def &(that: MethodDefs)(tn: TypeName, newdefns: Set[Str])(implicit raise: Raise): MethodDefs =
+      MethodDefs(
+        tn,
+        this.parents ::: that.parents,
+        mergeMap(this.decls, that.decls)(_ & _),
+        (this.defns.iterator ++ that.defns.iterator).toSeq.groupMap(_._1)(_._2).flatMap {
+          case _ -> Nil => die
+          case mn -> (defn :: Nil) => S(mn -> defn)
+          case mn -> defns if newdefns(mn) => N
+          case mn -> defns =>
+            implicit val prov: TypeProvenance = TypeProvenance(tn.toLoc, "inherited method declaration")
+            err(msg"An overriding method definition must be given when inheriting from multiple method definitions" -> tn.toLoc
+              :: msg"Definitions of method $mn inherited from:" -> N
+              :: defns.iterator.map(mt => msg"• ${mt.parents.head}" -> mt.prov.loco).toList)
+            S(mn -> MethodType(defns.head.level, N, tn :: Nil))
+        }
+      )
+    private def ++(that: MethodDefs): MethodDefs = {
```

Comment:
Not sure why this `private def` is needed. It's only used once in `propagate` and does not seem to make a lot of sense on its own.

Reply:
Addressed in dbf9e69d1f0b3293e36d97c6cf378bdfdb422fa8